src/Type.h

```diff
@@ -10,6 +10,128 @@
+    };
+    int32_t cpp_type_qualifiers; /// Bitset indicating which qualifiers are present on type
+
+    std::string name;
+
+    halide_cplusplus_type_name(CPPTypeType cpp_type_type, const std::string &name)
+        : cpp_type_type(cpp_type_type), cpp_type_qualifiers(0), name(name) {
+    }
+
+    bool operator==(const halide_cplusplus_type_name &rhs) const {
+         return cpp_type_type == rhs.cpp_type_type &&
+                cpp_type_qualifiers == rhs.cpp_type_qualifiers &&
+                name == rhs.name;
+    }
+
+    bool operator!=(const halide_cplusplus_type_name &rhs) const {
+        return !(*this == rhs);
+    }
+
+    bool operator<(const halide_cplusplus_type_name &rhs) const {
+         return cpp_type_type < rhs.cpp_type_type ||
+                (cpp_type_type == rhs.cpp_type_type &&
+                 (cpp_type_qualifiers < rhs.cpp_type_qualifiers ||
+                  (cpp_type_qualifiers == rhs.cpp_type_qualifiers &&
+                   name < rhs.name)));
+    }
+};
+   
+/** A structure to represent the fully scoped name of a C++ composite
+ * type for use in generating function signatures that use that type.
+ *
+ * This is intended to be a constexpr usable type, but we don't depend
+ * on C++11 yet. In C++14, it is possible this will be replaced with
+ * introspection/reflection facilities.
+ *
+ * TODO(zalman): Decide if this needs a field to indicate const/non-const.
+ */
+struct halide_handle_cplusplus_type {
+    halide_cplusplus_type_name inner_name;
+    std::vector<std::string> namespaces;
+    std::vector<halide_cplusplus_type_name> enclosing_types;
+    int32_t extra_indirection_levels; // Hack to handle void** on argv wrapper
+
+    halide_handle_cplusplus_type(const halide_cplusplus_type_name &inner_name,
+                                 const std::vector<std::string> &namespaces = std::vector<std::string>(),
+                                 const std::vector<halide_cplusplus_type_name> &enclosing_types = std::vector<halide_cplusplus_type_name>(),
+                                 int32_t extra_indirection_levels = 0)
+        : inner_name(inner_name), namespaces(namespaces), enclosing_types(enclosing_types), extra_indirection_levels(extra_indirection_levels) {
+    }
+};
+
+/** A type traits template to provide a halide_handle_cplusplus_type
+ * value from a C++ type.
+ *
+ * Note the type represented is implicitly a pointer.
+ * TODO(zalman): Figure out if we need to represent refs
+ *
+ * A NULL pointer of type halide_handle_traits represents "void *".
+ * This is chosen for compactness or representation as Type is a very
+ * widely used data structure.
+ */
+template<typename T>
+struct halide_handle_traits {
+    // NULL here means "void *". This trait must return a pointer to a
+    // global structure. I.e. it should never be freed.
+    static const halide_handle_cplusplus_type *type_info() { return nullptr; }
+};
+//@}
+
+template<>
+struct halide_handle_traits<int32_t *> {
```

Comment:
What's going on here with the specializations for int32_t and buffer_t? Does a user need to declare one of these specializations for every type they want to pass in as a handle?